sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala

```diff
@@ -650,4 +652,129 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
+
+            assert(getCatalogTable("spark_25474").stats.isEmpty)
+            val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
+            if (fallBackToHDFS) {
+              assert(relation.stats.sizeInBytes ===
+                CommandUtils.getSizeInBytesFallBackToHdfs(spark, getCatalogTable("spark_25474")))
+            } else {
+              assert(relation.stats.sizeInBytes === defaultSizeInBytes)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  test("Partitioned data source table stats should be cached") {
+    Seq(false, true).foreach { fallBackToHDFS =>
+      withSQLConf(SQLConf.ENABLE_FALL_BACK_TO_HDFS_FOR_STATS.key -> s"$fallBackToHDFS") {
+        withTempDir { dir =>
+          withTable("spark_25474") {
+            sql("CREATE TABLE spark_25474(a int, b int) USING parquet " +
+              s"PARTITIONED BY(a) LOCATION '${dir.toURI}'")
+            sql("INSERT INTO TABLE spark_25474 PARTITION(a=1) SELECT 2")
+
+            assert(getCatalogTable("spark_25474").stats.isEmpty)
+            val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
+            if (fallBackToHDFS) {
+              val dataSize =
+                CommandUtils.getSizeInBytesFallBackToHdfs(spark, getCatalogTable("spark_25474"))
+              assert(relation.stats.sizeInBytes === dataSize)
+
+              val qualifiedTableName =
+                QualifiedTableName(spark.sessionState.catalog.getCurrentDatabase, "spark_25474")
+              val logicalRelation = spark.sessionState.catalog.getCachedTable(qualifiedTableName)
+                .asInstanceOf[LogicalRelation]
+              assert(logicalRelation.catalogTable.get.stats.get.sizeInBytes === dataSize)
+            } else {
+              assert(relation.stats.sizeInBytes === conf.defaultSizeInBytes)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  test("External partitioned data source table does not support fallback to HDFS " +
+    "for size estimation") {
+    Seq(true).foreach { fallBackToHDFS =>
+      withSQLConf(SQLConf.ENABLE_FALL_BACK_TO_HDFS_FOR_STATS.key -> s"$fallBackToHDFS") {
+        withTempDir { dir =>
+          withTable("spark_25474") {
+            sql("CREATE TABLE spark_25474(a bigint, b bigint) USING parquet " +
+              s"PARTITIONED BY(a) LOCATION '${dir.toURI}'")
+
+            withTempDir { partitionDir =>
+              spark.range(5).write.mode(SaveMode.Overwrite).parquet(partitionDir.getCanonicalPath)
+              sql(s"ALTER TABLE spark_25474 ADD PARTITION (a=1) LOCATION '$partitionDir'")
+              assert(getCatalogTable("spark_25474").stats.isEmpty)
+              val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
+              assert(spark.table("spark_25474").count() === 5)
+              if (fallBackToHDFS) {
+                assert(relation.stats.sizeInBytes === 0)
```

Comment:
We have a serious issue here: the wrong stats may mislead Spark to broadcast a very large table and OOM.

I think we can only fallback to HDFS size for non-partitioned tables.

Reply:
+1 Partitioned table usually very large.

Reply:
@shahidki31 @dongjoon-hyun What do you think ?

Reply:
+1 for disabling for partitioned tables.
This is `master` branch status, isn't it? Previously, Spark returns `8EB` in this case safely.

Reply:
I think, non-partitioned data source table is already getting correct statistics. I am not sure, we need to support fallback to HDFS for size for non partitioned table.

Reply:
Yes. Non-partitioned data source table will get the correct statistics if table statistics are not available.

Reply:
Could we just revert https://github.com/apache/spark/commit/485ae6d1818e8756a86da38d6aefc8f1dbde49c2 ?

Reply:
I think, if we don't support fallback hdfs , we need to document saying non partitioned table join can be SMJ, even if the table size is below the broadcast threshold.

Reply:
+1 We need to document it. @shahidki31 Could you create a PR to revert 485ae6d?

Reply:
@wangyum . Why do we need to revert that? You can revert the functional part here and keep the test codes.

Reply:
For me, this PR already contains the reverting here(https://github.com/apache/spark/pull/24715#diff-fcb68cd3c7630f337ce9a3b479b6d0c4R70).

Reply:
Personally, I'm -1 for removing the existing test cases.

Reply:
@dongjoon-hyun It  is expensive to support partitioned table with external partitions. Please see [this test case](https://github.com/apache/spark/pull/24715/files#diff-8c27508821958acbe016862c9ab2f25fR770). Its data size is incorrect.

Related discussion:
https://github.com/apache/spark/pull/24715#discussion_r316143787
https://github.com/apache/spark/pull/24715#issuecomment-522414331

So we plan do not fallback to HDFS size for partitioned tables.

Reply:
> +1 We need to document it.
@wangyum May be in this PR only you can do both, if doesn't support fallback config.